src/library/scala/util/Using.scala

```
@@ -148,10 +148,12 @@ object Using {
   def resource[R: Resource, A](resource: R)(body: R => A): A = {
     if (resource == null) throw new NullPointerException("null resource")
 
-    @inline def safeAddSuppressed(t: Throwable, suppressed: Throwable): Unit = {
-      // don't `addSuppressed` to something which is a `ControlThrowable`
-      if (!t.isInstanceOf[ControlThrowable]) t.addSuppressed(suppressed)
-    }
+    @inline def safeAddSuppressed(t: Throwable, suppressed: Throwable): Throwable =
+      if (t.isInstanceOf[ControlThrowable] && !suppressed.isInstanceOf[ControlThrowable]) {
+        suppressed.addSuppressed(t) ; suppressed
+      } else {
+        t.addSuppressed(suppressed) ; t
+      }
```

I'm not sure I would consider this new behavior desirable. `ControlThrowable` (e.g. `NonLocalReturnControl`) *needs* to propagate, even if it doesn't pick up suppression information. For example:

```scala
val totalLines =
  (1 to 10)
    .map(_.toString)
    .foldLeft(0) { (count, i) =>
      Using(new BufferedReader(new FileReader(s"$i.txt"))) { br =>
        val lines = br.lines.asScala
        if (lines.isEmpty) return -1
        count + lines.length
      } getOrElse count
    }
```

If for some reason closing the file fails, the intended `-1` result may not get returned (depending on whether or not the compiler was able to elide the `NonLocalReturnControl`).

For reference, I consciously decided to have `ControlThrowable` drop suppressed exceptions, knowing they would vanish, when I wrote it.

---

I haven't played around with it yet, but my intuition was that if I try to produce a value but an exception is thrown, then the exception is thrown. If I wanted to encode a value and also collect all the errors along the way, I'd be doing something else. Ignoring exceptions doesn't sound right (and in particular, assuming exceptions from closing things don't matter). But it also seems like `ControlThrowable` shouldn't suppress: or is that wrong?

---

I used WIP label to indicate I was gathering feedback, but of course I still value feedback from you who has gone down this road already. I don't have firm opinions in these matters at all. I'll try to work up some experience or maybe some tests.

---

Sorry it's taken me so long to reply to this.

If you take the example I used and just replace `Using` with `Using.resource`, you actually want the opposite behaviour (i.e. not throwing the `NonLocalReturnControl`. Consequently, I don't think there's a good solution that "makes sense" or works universally. I think the simplest option (not giving it any special treatment) is probably best.

The only reason there was any special treatment at all for `ControlThrowable` in the first place was so that cached instances (e.g. the ones used by `scala.util.control.Breaks`) would not leak memory. If you add a suppressed exception to a cached instance, it will never get GCed, and if that happens repeatedly, you could end up leaking a very large amount of memory.

For non-cached instances (e.g. `NonLocalReturnControl`), there's no benefit to suppressing the exception (it will never be seen), but it also doesn't hurt, because they are single-use, and get GCed themselves in short order.

---

A possible way to prevent exceptions from vanishing would be if the exception doesn't support suppression, throw the exception that would have been added as suppressed on a different thread. I'm not really a fan of this idea though.

---

I see the `Using` use case is in flux, or in flummox, so I'll back out the last commit.